plugins/hive-plugin/src/main/java/edu/berkeley/ground/plugins/hive/GroundDatabase.java

```diff
@@ -43,311 +45,171 @@
+        // number of databases
+        if (metaDatabaseNodeVersion == null) {
+            try {
+                Database metaDatabase = new Database(DATABASE_PARENT_NODE, "metanode", "locationurl",
+                        new HashMap<String, String>());
+                metaDatabaseNodeVersion = createDatabaseNodeVersion(metaDatabase, EntityState.ACTIVE.name());
+            } catch (InvalidObjectException | MetaException ex) {
+                LOG.error("error creating metadatabase: {}", ex);
+            }
+        }
     }
-  }
 
-  Structure getNodeStructure(String dbName) throws GroundException {
-    try {
-      Node node = this.getNode(dbName);
-      return groundReadWrite.getStructureFactory().retrieveFromDatabase(dbName);
-    } catch (GroundException e) {
-      LOG.error("Unable to fetch database node structure");
-      throw e;
+    Node getNode(String dbName, Map<String, Tag> tagMap) throws GroundException {
+        LOG.debug("Fetching database node: " + dbName);
+        Node node = this.groundReadWrite.getGroundReadWriteNodeResource().getNode(dbName);
+        if (node == null) {
+            return this.groundReadWrite.getGroundReadWriteNodeResource().createNode(dbName, tagMap);
+        }
+        return node;
     }
-  }
 
-  Edge getEdge(NodeVersion nodeVersion) throws GroundException {
-    String edgeId = "" + nodeVersion.getNodeId();
-    try {
-      LOG.debug("Fetching database table edge: " + edgeId);
-      return groundReadWrite.getEdgeFactory().retrieveFromDatabase(edgeId);
-    } catch (GroundException e) {
-      LOG.debug("Not found - Creating database table edge: " + edgeId);
-      Edge edge = groundReadWrite.getEdgeFactory().create(edgeId, new HashMap<>());
-      Structure edgeStruct = groundReadWrite.getStructureFactory().create(edge.getName(), new HashMap<>());
-      return edge;
+    Database getDatabase(String dbName) throws GroundException {
+        NodeVersion latestVersion = getDatabaseNodeVersion(dbName);
+        LOG.info("database versions id: {}", latestVersion.getNodeId());
+        LOG.info("database versions structureVersion: {}", latestVersion.getStructureVersionId());
+        Map<String, Tag> dbTag = latestVersion.getTags();
+        return PluginUtil.fromJson((String) dbTag.get(dbName).getValue(), Database.class);
     }
-  }
 
-  Structure getEdgeStructure(NodeVersion nodeVersion) throws GroundException {
-    try {
-      LOG.debug("Fetching database table edge structure: " + nodeVersion.getNodeId());
-      Edge edge = this.getEdge(nodeVersion);
-      return groundReadWrite.getStructureFactory().retrieveFromDatabase(edge.getName());
-    } catch (GroundException e) {
-      LOG.debug("Not found - database table edge structure: " + nodeVersion.getNodeId());
-      throw e;
+    NodeVersion getDatabaseNodeVersion(String dbName) throws GroundException {
+        List<Long> versions = (List<Long>) PluginUtil.getLatestVersions(dbName, "nodes");
```

Comment:
I mentioned this before. by default Jackson returns <Int> array. If the Cast is removed, you will get a class cast exception